pkg/cloudmap/utils.go

```diff
@@ -24,3 +37,61 @@ func ShouldPodBeInEndpoints(pod *corev1.Pod) bool {
 		return true
 	}
 }
+
+func AwaitOperationSuccess(timeoutSeconds time.Duration, tickerSeconds time.Duration, f func() string) string {
+	timeout := time.After(timeoutSeconds * time.Second)
+	ticker := time.Tick(tickerSeconds * time.Second)
+	operationStatus := make(chan string, 1)
+	go func() {
+		for {
+			select {
+			case <-timeout:
+				operationStatus <- TIMEOUT
+				return
+			case <-ticker:
+				status := f()
+				if status == SUCCESS {
+					operationStatus <- status
+					return
+				}
+			}
+		}
+	}()
+	return <-operationStatus
+}
```

Comment:
i dont think your function will timeout appropriately if the function passed in `f()` exceeds the timeout duration. for example if you run this test which is a modification of what you have:

```golang
func TestAwaitOperationSuccess(t *testing.T) {
	tests := []struct {
		name    string
		timeout time.Duration
		tick    time.Duration
		want    string
	}{
		{
			name:    "Await function should succeed",
			timeout: 10 * time.Second,
			tick:    1 * time.Second,
			want:    SUCCESS,
		},
		{
			name:    "Await function should timeout",
			timeout: 2 * time.Second,
			tick:    1 * time.Second,
			want:    TIMEOUT,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			status := AwaitOperationSuccess(tt.timeout, tt.tick, func() string {
				time.Sleep(3 * time.Second)
				return SUCCESS
			})

			assert.Equal(t, tt.want, status)
		})
	}
}
```
output:
```
=== RUN   TestAwaitOperationSuccess
=== RUN   TestAwaitOperationSuccess/Await_function_should_succeed
=== RUN   TestAwaitOperationSuccess/Await_function_should_timeout
    utils_test.go:37: 
        	Error Trace:	utils_test.go:37
        	Error:      	Not equal: 
        	            	expected: "TIMEOUT"
        	            	actual  : "SUCCESS"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-TIMEOUT
        	            	+SUCCESS
        	Test:       	TestAwaitOperationSuccess/Await_function_should_timeout
--- FAIL: TestAwaitOperationSuccess (8.00s)
    --- PASS: TestAwaitOperationSuccess/Await_function_should_succeed (4.00s)
    --- FAIL: TestAwaitOperationSuccess/Await_function_should_timeout (4.00s)
FAIL
FAIL	command-line-arguments	8.670s
FAIL
```

indicating that the timeout of 2 seconds is not respected bc the function passed in sleeps for 3 seconds. there are a couple of ways to do this, but i believe what you want is to create two separate goroutines that are competing for completion and receive/return the faster of the two. something like the following:

```golang
func AwaitOperationSuccess(timeout time.Duration, tick time.Duration, f func() string) string {
	res := make(chan string, 1)

	go func() {
		time.Sleep(timeout) // or <-time.After(timeout) doesnt really matter
		res <- TIMEOUT
	}()

	go func() {
		for _ = range time.Tick(tick) {
			status := f()
			if status == SUCCESS {
				res <- SUCCESS
                                 return
			}
		}
	}()

	return <-res
}
```

Reply:
what is the expectation for the passed-in function? its a bit odd that we are creating custom output strings in the utils file that vary from the usages in say `resource_manager.go` but we have a line ```status == SUCCESS``` where we expect the output of the passed-in function to be a certain value. i see from some digging that the output of the usage in `resource_manager.go` can be these enums https://github.com/aws/aws-sdk-go/blob/main/service/servicediscovery/api.go#L6637

Reply:
For the strings (SUCCESS, TIMEOUT, FAIL), I created them in the utils.go file so that I could use them in the resource_manager.go, utils.go, and (eventually) the related integration tests. I was thinking I could use those enums, but there isn't a TIMEOUT one that would indicate that the related async operation will possibly succeed but is taking too long. I also thought it looked better to define those as constants since they would be used frequently.